### PR TITLE
Feat(web-twig): Introduce styleProps, deprecate class in Alert #DS-646

### PR DESCRIPTION
**Pull request by crishpeen**

### CODE REVIEW

File: packages/web-twig/src/Twig/PropsExtension.php [hunk 1]
```diff
@@ -91,4 +96,29 @@ public function renderClassProp(Environment $environment, array $classNames): st
...
         ]);
     }
+
+    /**
+     * @param array<string, mixed> $props
+     */
+    public function renderStyleProp(Environment $environment, array $props): string
+    {
+        return $environment->render('@partials/styleProp.twig', [
+            'props' => $props,
```
Comment (literat) [inline packages/web-twig/src/Twig/PropsExtension.php:106]:
```suggestion
            'style' => $props['style'],
```

File: packages/web-twig/src/Twig/PropsExtension.php [hunk 2]
```diff
@@ -91,4 +96,29 @@ public function renderClassProp(Environment $environment, array $classNames): st
...
+    }
+
+    /**
+     * @param array<string, mixed> $props
+     * @return array<string, mixed>
+     */
+    public function useStyleProps(array $props): array
+    {
+        $data = [];
+        if (isset($props['class'])) {
```
Comment (literat) [inline packages/web-twig/src/Twig/PropsExtension.php:117]:
```suggestion
        if (array_key_exists('class', $props)) {
```

https://www.php.net/manual/en/function.array-key-exists

Just the existence of the key is enough to raise the exception.

File: packages/web-twig/src/Twig/PropsExtension.php [hunk 3]
```diff
@@ -91,4 +96,29 @@ public function renderClassProp(Environment $environment, array $classNames): st
...
+            trigger_error('Use UNSAFE_className instead of class', E_USER_DEPRECATED);
```
Comment (literat) [inline packages/web-twig/src/Twig/PropsExtension.php:118]: Please use a full deprecation message. Same as in other places.

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-twig/src/Resources/components/Alert/README.md [hunk 4]
```diff
@@ -29,13 +29,14 @@ Without lexer:
...
-| `iconName`    | `string`                                     | `null`    | no       | Icon used in Alert         |
-| `isCentered`  | `bool`                                       | `false`   | no       | If true, Alert is centered |
+| Prop name          | Type                                         | Default   | Required | Description                |
+| ------------------ | -------------------------------------------- | --------- | -------- | -------------------------- |
+| `color`            | [Emotion Color dictionary][dictionary-color] | `success` | no       | Color variant              |
+| `elementType`      | `string`                                     | `div`     | no       | HTML tag to render         |
+| `iconName`         | `string`                                     | `null`    | no       | Icon used in Alert         |
+| `isCentered`       | `bool`                                       | `false`   | no       | If true, Alert is centered |
+| `UNSAFE_className` | `string`                                     | -         | no       | Custom root class name     |
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Alert/README.md:38]: Maybe it would be better to only add a note and a link here to another page with documentation about escape hatches. See https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web-react/README.md#escape-hatches.

E.g. what about moving this part to root readme or somewhere else because it should be the same in both packages?

File: packages/web-twig/src/Twig/PropsExtension.php [hunk 5]
```diff
@@ -29,6 +29,11 @@ public function getFunctions(): array
                 'needs_environment' => true,
                 'is_safe' => ['html'],
             ]),
+            new TwigFunction('styleProp', [$this, 'renderStyleProp'], [
+                'needs_environment' => true,
+                'is_safe' => ['html'],
```
Comment (literat) [inline packages/web-twig/src/Twig/PropsExtension.php:34]: This option disables escaping, see https://twig.symfony.com/doc/2.x/advanced.html#automatic-escaping

Please remove this line and add some examples with custom styles to ensure that everything is safely rendered. Thanks.

For more information see our discussion on Slack: https://almamedia.slack.com/archives/C0502FGLG3G/p1681476431357859 or contact @jirinovak

**Review by literat — CHANGES_REQUESTED** [review]
Please fix the comment with the escaping, thanks :-)

File: packages/web-twig/src/Twig/PropsExtension.php (hunk 5 above)
Comment (crishpeen) [reply packages/web-twig/src/Twig/PropsExtension.php:34]: https://slack-files.com/T0325RBAD-F053G5D059T-2af92bb610 I have to keep it here as it breaks quotes in the embedded twig file.

**Review by literat — APPROVED** [review]
Nice and clean ❤️  Thanks.

File: packages/web-twig/src/Resources/components/Alert/stories/AlertCentered.twig [hunk 6]
```diff
@@ -1,17 +1,17 @@
-<Alert color="success" isCentered class="mb-600">
+<Alert color="success" isCentered UNSAFE_className="mb-600">
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Alert/stories/AlertCentered.twig:1]: @crishpeen @literat I think we soon need to find a way how to apply utility classes because there is nothing unsafe about adding a margin. As stated in the docs ⬆️ , `UNSAFE` props should be reserved for edge cases, not for everyday use. I'm filing it for our weekly discussion.

File: packages/web-twig/tests/ComponentTest.php [hunk 7]
```diff
@@ -134,6 +134,40 @@ class="Button Button--primary Button--medium"
         );
     }
 
+    public function testShouldRenderComponentWithUnsafeCaseNameAndStyle(): void
+    {
+        $html = $this->twig->render('test_component_classname_style.twig');
+        $html = $this->removeWhitespace($html);
+
+        $this->assertEquals(
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-twig/tests/ComponentTest.php:142]: I refactored them. Now outputs are in HTML files and we basically compare the twig and the HTML files.

File: packages/web-twig/src/Twig/PropsExtension.php [hunk 8]
```diff
@@ -91,4 +96,30 @@ public function renderClassProp(Environment $environment, array $classNames): st
...
+        return $environment->render('@partials/styleProp.twig', [
+            'style' => $props['style'] ?? null,
+        ]);
+    }
+
+    /**
+     * @param array<string, mixed> $props
+     * @return array<string, mixed>
+     */
+    public function useStyleProps(array $props): array
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/src/Twig/PropsExtension.php:114]: @adamkudrna Well, it is still a function. Returns the same value and logic is also the same. You are right, that the context is different but the concept is similar or close. Also, we are using this the same way as we did in React. So it makes sense to me to name it the same way.

Please, feel free to suggest better naming here. As we are trying a little bit to simulate React component I consider this to be fine at this moment. 🤷

[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-twig/src/Twig/PropsExtension.php:114]: I am closing this for now. We might improve it in the future.

**Review by literat — APPROVED** [review]
